tests/test_pipelines_image_segmentation.py

```diff
@@ -0,0 +1,282 @@
+# Copyright 2021 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import hashlib
+import unittest
+
+from transformers import (
+    MODEL_FOR_IMAGE_SEGMENTATION_MAPPING,
+    AutoFeatureExtractor,
+    AutoModelForImageSegmentation,
+    ImageSegmentationPipeline,
+    is_vision_available,
+    pipeline,
+)
+from transformers.testing_utils import (
+    is_pipeline_test,
+    nested_simplify,
+    require_datasets,
+    require_tf,
+    require_timm,
+    require_torch,
+    require_vision,
+    slow,
+)
+
+from .test_pipelines_common import ANY, PipelineTestCaseMeta
+
+
+if is_vision_available():
+    from PIL import Image
+else:
+
+    class Image:
+        @staticmethod
+        def open(*args, **kwargs):
+            pass
+
+
+@require_vision
+@require_timm
+@require_torch
+@is_pipeline_test
+class ImageSegmentationPipelineTests(unittest.TestCase, metaclass=PipelineTestCaseMeta):
+    model_mapping = MODEL_FOR_IMAGE_SEGMENTATION_MAPPING
+
+    @require_datasets
+    def run_pipeline_test(self, model, tokenizer, feature_extractor):
+        def test_outputs(outputs):
+            png_string, segments_info = outputs["png_string"], outputs["segments_info"]
+            self.assertEqual(png_string, ANY(str))
+            self.assertGreater(len(segments_info), 0)
+            for segment_info in segments_info:
+                self.assertEqual(
+                    segment_info,
+                    {
+                        "id": ANY(int),
+                        "score": ANY(float),
+                        "label": ANY(str),
+                    },
+                )
+
+        image_segmenter = ImageSegmentationPipeline(model=model, feature_extractor=feature_extractor)
+        outputs = image_segmenter("./tests/fixtures/tests_samples/COCO/000000039769.png", threshold=0.0)
+        test_outputs(outputs)
+
+        subtasks = [None, "panoptic", "instance", "semantic"]
+        for subtask in subtasks:
```

Comment:
Never use for loops in tests. Or very exceptionally.
Never use inner functions in tests.

Both really hurt the main asset of a test, which is readability. A reader needs to be able to read a single test and know if it's valid or not, with as little as possible surrounding context.

Here, reader needs to understand what variable is sent to `test_outputs`, understand the function itself, then reason if variable was supposed to be A, [A] or something else to assess if the test is correct.


```python
outputs = image_segmenter("./tests/fixtures/tests_samples/COCO/000000039769.png", threshold=0.0)
self.assertEqual(nested_simply(outputs, {.....})

# Easy to read that if input is list, output is list.
outputs = image_segmenter(["./tests/fixtures/tests_samples/COCO/000000039769.png"], threshold=0.0)
self.assertEqual(nested_simply(outputs, [{.....}])

outputs = image_segmenter("./tests/fixtures/tests_samples/COCO/000000039769.png", threshold=0.0, subtask="panoptic")
self.assertEqual(nested_simply(outputs, {.....})

....
```